### PR TITLE
Capture UNIX domain socket packet

### PR DESCRIPTION
**Pull request by Rtoax**
I commit it to "tools/" before(https://github.com/iovisor/bcc/pull/3598), push it here "examples/tracing" better.

1. start up server
[rongtao@localhost tools]$ nc --unixsock /work/workspace/undump3.sock -l

2. start undump.py 
[root@localhost study]# ./undump.py -p $(pidof nc)
Tracing UNIX socket packets ... Hit Ctrl-C to end
PID    COMM

3. startup UNIX  socket client
[rongtao@localhost ~]$ nc --unixsock /work/workspace/undump3.sock

4. input some words
[rongtao@localhost ~]$ nc --unixsock /work/workspace/undump3.sock
Hello, World

5. capture the packet
----------------48 65 6c 6c 6f 2c 20 57 6f 72 6c 64 0a 
----------------recv 13 bytes
153784 nc

### CODE REVIEW

File: examples/tracing/undump.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+        sys.stdout.flush()
+        if (i+1)%16 == 0:
+            print("")
+            print("----------------", end="")
+    print("\n----------------recv %d bytes" % event.recv_len)
+    
+# initialize BPF
+b = BPF(text=bpf_text)
+b.attach_kprobe(event="unix_stream_recvmsg", fn_name="trace_stream_entry")
+b.attach_kprobe(event="unix_stream_read_actor", fn_name="trace_unix_stream_read_actor")
```
Comment (chenhengqi) [inline examples/tracing/undump.py:179]: I am curious why you need to trace two functions to capture packets. 

I think `unix_stream_recvmsg` should be enough.

File: examples/tracing/undump.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+print("Tracing UNIX socket packets ... Hit Ctrl-C to end")
+
+# header
+if args.timestamp:
+    print("%-9s" % ("TIME(s)"), end="")
+
+print("%-6s %-12s" % ("PID", "COMM"), end="")
+print()
+
+print()
```
Comment (chenhengqi) [inline examples/tracing/undump.py:190]: extra newline

File: examples/tracing/undump.py (hunk 1 above)
Comment (Rtoax) [reply examples/tracing/undump.py:179]: “unix_stream_recvmsg” can not get sk_buff directly, "trace_unix_stream_read_actor" is easier. but must use "unix_stream_recvmsg" get  COMM.

Comment (chenhengqi) [reply examples/tracing/undump.py:179]: Do you mean that `trace_unix_stream_read_actor` is executed under softirq context ?

Comment (Rtoax) [reply examples/tracing/undump.py:179]: No， `unix_stream_read_actor` is easy to get sk_buff, when `recv msg` happened. 
function call stack like:
```
unix_stream_recvmsg() -----> can not get sk_buff, or hardly, but use it to get/filter COMM and  PID
   ...
      unix_stream_read_actor() -----> get sk_buff is easier than unix_stream_recvmsg.
```

Comment (chenhengqi) [reply examples/tracing/undump.py:179]: So why not just use `trace_unix_stream_read_actor` ?

File: examples/tracing/undump.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    data4.sock_flags = sock->flags;
+
+    data4.msg_namelen = msg->msg_namelen;
+    
+    bpf_get_current_comm(&data4.task, sizeof(data4.task));
+    
+    struct unix_sock *unsock = (struct unix_sock *)sock->sk;
+    data4.unix_sock_path = (char *)unsock->path.dentry->d_name.name;
+    
+    stream_recvmsg_events.perf_submit(ctx, &data4, sizeof(data4));
```
Comment (chenhengqi) [inline examples/tracing/undump.py:113]: You fill 9 fields here, but userspace only print 3 of them ?

File: examples/tracing/undump.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+int trace_unix_stream_read_actor(struct pt_regs *ctx)
+{
+    u32 zero = 0;
+    int ret = PT_REGS_RC(ctx);
+    u64 pid_tgid = bpf_get_current_pid_tgid();
+    u32 pid = pid_tgid >> 32;
+    u32 tid = pid_tgid;
+
+    FILTER_PID
```
Comment (chenhengqi) [inline examples/tracing/undump.py:126]: You have `FILTER_PID` here which is supposed to run under process context, so `bpf_get_current_comm` should also work, right ?

File: examples/tracing/undump.py (hunk 3 above)
Comment (Rtoax) [reply examples/tracing/undump.py:113]: You are right, i'll modify it

File: examples/tracing/undump.py (hunk 4 above)
Comment (Rtoax) [reply examples/tracing/undump.py:126]: You are right, i'll modify it

Comment (davemarchevsky) [reply examples/tracing/undump.py:76]: I think this is good to go, pending this comment being addressed. @chenhengqi do you think it's fine to merge w/o another iteration?

Comment (chenhengqi) [reply examples/tracing/undump.py:76]: Yeah, since this tool is served as an example not a BCC tool, I think it is good to go.

**Comment by davemarchevsky** [comment]
[buildbot, test this please]

**Comment by davemarchevsky** [comment]
whoops, forgot to ask you to squash. Oh well. Thanks for the contribution!

**Pull request merged**